### ISSUE TITLE
Avoid `getBean` method in `Infinispan*CacheManagerChecker` classes

### ISSUE THREAD

**Issue by slaskawi** [infinispan/infinispan-spring-boot#38]
Currently `InfinispanRemoteCacheManagerChecker#hasConfigurer` and `InfinispanRemoteCacheManagerChecker#hasConfiguration` methods use `conditionContext#getBeanFactory()#getBean` methods. This causes problems with injecting `Environment` or processing `@Value`s in client's configuration. Here's an example:

```
@Configuration
public class HotRodConfiguration {
    @Value("#{environment.SERVER_LIST}")
    private String serverList;

    @Bean
    public InfinispanRemoteConfigurer configuration() {
        return () -> new ConfigurationBuilder()
              .addServers(serverList)
              .build();
    }
}
```

In the example above the `serverList` field will expand to `null` and this warning will be emitted to the console:

```
2017-07-20 20:11:18.989  WARN 7424 --- [           main] o.s.c.a.ConfigurationClassPostProcessor  : Cannot enhance @Configuration bean definition 'hotRodConfiguration' since its singleton instance has been created too early. The typical cause is a non-static @Bean method with a BeanDefinitionRegistryPostProcessor return type: Consider declaring such methods as 'static'.
```

The root cause is the `getBean` method evaluated inside a class which implements `Condition`. It forces Spring to eagerly create `HotRodConfiguration` bean (since it contains `InfinispanRemoteConfigurer`) it has no chance to process the configuration later on.

**Comment by slaskawi** [comment]
Of course There's an easy work around for this specific example: use `System#getenv` :D

**Comment by slaskawi** [comment]
//cc @lburgazzoli Just in case you were interested in this...

**Comment by lburgazzoli** [comment]
Would that work if we move the bean check on a condition extending AnyNestedCondition ? 
Like:

```java
public class ConditionalOnAnyBean extends AnyNestedCondition {
    public ConditionalOnAnyBean() {
        super(ConfigurationPhase.REGISTER_BEAN);
    }

    @ConditionalOnBean(InfinispanRemoteConfigurer.class)
    static class OnRemoteConfigurer{
    }

    @ConditionalOnBean(Configuration.class)
    static class OnConfiguration {
    }
}
```

**Comment by slaskawi** [comment]
I don't think so. The root cause is [this construct conditionContext.getBeanFactory().getBean(InfinispanRemoteConfigurer.class)](https://github.com/infinispan/infinispan-spring-boot/blob/master/infinispan-spring-boot-starter-remote/src/main/java/org/infinispan/spring/starter/remote/InfinispanRemoteCacheManagerChecker.java#L26). But it should be possible to go through the BeanDefinitions (created before instantiating beans) and check if bean of a class `InfinispanRemoteConfigurer` will be created or not. 

Unfortunately this requires playing a bit with `getBeanFactory().getBeanDefinition` and `getBeanFactory().getBeanDefinitionNames` methods.

**Comment by lburgazzoli** [comment]
for my understanding, according to the doc @ConditionalOnBean should check if the bean class/name is already contained int he BeanFactory so it should effectively do the same as inspecting the bean definitions, am I missing something ?

**Comment by slaskawi** [comment]
Got you now. Interesting idea. It's worth to check it out.

**Comment by lburgazzoli** [comment]
Is there an unit test for this ?

**Comment by slaskawi** [comment]
Unfortunately not. But it would be trivial - just create a `@Configuration` with `InfinispanRemoteConfigurer` and inject `ApplicationContext` or `Environment` to this configuration class.

BTW - I'll be travelling for the next a couple of days, so I will be able look into it in a week. If you have a free slot (and if this issue is interesting to you), feel free to implement it.

**Comment by lburgazzoli** [comment]
I've opened a PR, it may need a little beet of cleanup but it appears to work.

**Comment by lburgazzoli** [comment]
Fixed by a1f7c884b79493086c63a8dcafead069e991b01c
